x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/string/InsertFunctionProcessor.java

```diff
@@ -0,0 +1,130 @@
+import org.elasticsearch.xpack.sql.expression.function.scalar.processor.runtime.Processor;
+
+import java.io.IOException;
+import java.util.Objects;
+
+public class InsertFunctionProcessor implements Processor {
+
+    private final Processor source, start, length, replacement;
+    public static final String NAME = "ins";
+
+    public InsertFunctionProcessor(Processor source, Processor start, Processor length, Processor replacement) {
+        this.source = source;
+        this.start = start;
+        this.length = length;
+        this.replacement = replacement;
+    }
+
+    public InsertFunctionProcessor(StreamInput in) throws IOException {
+        source = in.readNamedWriteable(Processor.class);
+        start = in.readNamedWriteable(Processor.class);
+        length = in.readNamedWriteable(Processor.class);
+        replacement = in.readNamedWriteable(Processor.class);
+    }
+
+    @Override
+    public final void writeTo(StreamOutput out) throws IOException {
+        out.writeNamedWriteable(source());
+        out.writeNamedWriteable(start());
+        out.writeNamedWriteable(length());
+        out.writeNamedWriteable(replacement());
+    }
+
+    @Override
+    public Object process(Object input) {
+        return doProcess(source().process(input), start().process(input), length().process(input), replacement().process(input));
+    }
+
+    public static Object doProcess(Object source, Object start, Object length, Object replacement) {
+        if (source == null) {
+            return null;
+        }
+        if (!(source instanceof String || source instanceof Character)) {
+            throw new SqlIllegalArgumentException("A string/char is required; received [{}]", source);
+        }
+        if (replacement == null) {
+            return source;
+        }
+        if (!(replacement instanceof String || replacement instanceof Character)) {
+            throw new SqlIllegalArgumentException("A string/char is required; received [{}]", replacement);
+        }
+        if (start == null || length == null) {
+            return source;
+        }
+        if (!(start instanceof Number)) {
+            throw new SqlIllegalArgumentException("A number is required; received [{}]", start);
+        }
+        if (!(length instanceof Number)) {
+            throw new SqlIllegalArgumentException("A number is required; received [{}]", length);
+        }
+        if (((Number) length).intValue() < 0) {
+            throw new SqlIllegalArgumentException("A positive number is required for [length]; received [{}]", length);
+        }
+
+        int startInt = ((Number) start).intValue() - 1;
+        int realStart = startInt < 0 ? 0 : startInt;
+        StringBuilder sb = new StringBuilder(source instanceof String ? (String) source : ((Character) source).toString());
+        String replString = (replacement instanceof String ? (String) replacement : ((Character) replacement).toString());
+        
+        if (startInt > sb.length()) {
```

Comment:
I think it is worth double checking the bounds here because you already subtract one from it. It is sneaky enough that we need to have a tests case for it.

Reply:
Also maybe check this before building `sb` and stuff.

Reply:
@nik9000 good point about building `sb` AFTER the possible `return`.
But I'm failing to see the issue with `startInt`s comparison with the length. Indeed, `startInt` can be negative, but when it happens `realStart` (the one used with StringBuilder) will be 0 and it shouldn't be an issue I think...